meshcap update

DeniseWorthen opened the pull request:
# Pull Request Summary

Update mesh cap with changes needed for eventual merge to develop.

## Description

- Allows users of the mesh cap to set runtime options for custom restart and history file names and to create netcdf output for gridded fields at run time. For UWM, these can be set in the nems.configure using the settings

```
  gridded_netcdfout = true
  user_sets_histname = true
  user_sets_restname = true
```

History file names will then be ``casename.ww3.hi.YYYY-MM-DD-SSSS.nc`` and ``casename.ww3.r.YYYY-MM-DD-SSSS`` where ``casename`` is set from nems.configure for each application (e.g ``ufs.hafs``)

- Replaces previous ``set_shel_inp`` and ``read_shel_inp`` routines with the applicable section from ``ww3_shel.F90`` to read either a inp or nml file. When used in UFS, the only required fields in the nml file are the list of forcing fields, the desired output fields, the name of the points file (if desired) and the stride for history and restart files. For example,

```
&input_nml
  input%forcing%winds = 'T'
  input%forcing%currents  = 'F'
  input%forcing%ice_conc  = 'F'
/
&output_type_nml
  type%field%list     = 'WND HS FP DP PHS PTP PDIR'
  type%point%file     = 'points.list'
/

&output_date_nml
  date%field%outffile  = '1'
  date%field%stride    = '3600'
  date%point%stride    = '3600'
  date%restart2%stride = '3600'
/
```

- Enables a set of logical flags to replace a small sub-set of current ifdefs to control program execution as a first step towards to removing ifdefs and using runtime options. These flags are created in ``wav_shr_flags`` and  implemented at this time primarily in ``w3wavemd`` and ``w3initmd`` for the various debug messaging and tracing ifdefs. This change was primarily accomplished via python script developed by @mvertens. The script also leverages emacs to use standard F90 if-then-else indenting. This introduces a large number of white-space changes. Neglecting whitespace changes can be view using the ?w=1 modification to the file difference [url](https://github.com/NOAA-EMC/WW3/pull/757/files?w=1)

- Streamlines some sections of code which, after implementing logical flags, were found contain a serial string of the same ifdef (eg. SR W3SCAT)

- Removes ``wav_tauice`` fields which are unused when coupling WW3 to CICE6.

<!--
Provide a detailed description of what this PR does.
What bug does it fix, or what feature does it add?
Is a change of answers expected from this PR?

Please also include the following information: 
* Add any suggestions for a reviewer 
* Mention any labels that should be added:  _bug_, _documentation_, _enhancement_, _new feature_
* Are answer changes expected from this PR? Please describe the changes and the reason why in addition to which of the following labels would apply: _mod_def change_, _out_grd change_, _out_pnt change_, _restart file change_, _Regression test_ 
-->

### Issue(s) addressed
- https://github.com/NOAA-EMC/WW3/issues/679
- https://github.com/NOAA-EMC/WW3/issues/674
- https://github.com/NOAA-EMC/WW3/issues/673
- https://github.com/NOAA-EMC/WW3/issues/676
- 
### Commit Message

Update mesh cap for custom user filenames, netcdf gridded output and logical control of some ifdef options.
 
co-authors : @mvertens 

### Testing

All UFS-WM baselines were run using hera.intel and hera.gnu and were B4B.

DeniseWorthen committed:
collect common fields for cesm and ufs

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/w3capbs1

DeniseWorthen committed:
remove extra odat declaration

DeniseWorthen committed:
add stokes export

* global winds retrieved w/ correct name
* add stokes variables to exports

DeniseWorthen committed:
remove debugging line

DeniseWorthen committed:
change fill value for stokes

DeniseWorthen committed:
tidy up

DeniseWorthen committed:
initial repro w/ stokes export on

* z0 not repro but value exceeds max for atm use so
no impact

DeniseWorthen committed:
Merge remote-tracking branch 'origin/dev/ufs-weather-model' into feature/w3capbs1

DeniseWorthen committed:
restart repro with temp debugging code

* add reset of charn for activated/deactivated ice points
* ran both 6h and 12h restart cases, both passed
* use mod_def.mx100, created using glo_1deg grid.inp except
for turning off k-shifting (broke repro!)
* temp debugging code to be removed

DeniseWorthen committed:
remove debugging switches

DeniseWorthen committed:
remove un-used/needed items for ufs

* control restarts using odat ?? make here 3600s
* remove/comment out items that have no impact on ufs (why?)
* tidy up

DeniseWorthen committed:
reduce restart freq to 3h

DeniseWorthen committed:
fix inflags

* turning these off gives all zeros in exported fields

DeniseWorthen committed:
fix inflags, momentum

* set inflags(2:4) true, others false. Treat momentum (if used)
like wind in filling global data

DeniseWorthen committed:
bring in Matts debug changes 0096df6e

DeniseWorthen committed:
update cmplr.env

DeniseWorthen committed:
tidy up

DeniseWorthen committed:
fix errors shown by debug mode

DeniseWorthen committed:
Merge branch 'feature/w3capbs1_debug' into feature/w3capbs1

DeniseWorthen committed:
revert build files

DeniseWorthen committed:
add changes from w3capbs1 branch

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/ww3capbuild

DeniseWorthen committed:
add back wcoss compile option and tidy up

DeniseWorthen committed:
replace wcoss compile options

DeniseWorthen committed:
another esmf/Makefile update

DeniseWorthen committed:
remove state_diagnose at realize, tidy up

* state_diagnose at realize failed in debug mode because of
sum() of uninitialized variables
* remove debug switches
* remove extra 'HK' print statements

DeniseWorthen committed:
more cleanup,alignment w/ mcapbuild,capbs1 branches

DeniseWorthen committed:
c96,c192,c384 tests all have b4b coupler restarts

DeniseWorthen committed:
tidy up

DeniseWorthen committed:
fix indent, remove netcdf q/dirty line

DeniseWorthen committed:
initial commit of wav_shel_inp

* model completes 6hrs in debug mode and writes out_pnt.points
and out_grd.mx100

DeniseWorthen committed:
prevent over-writing by local cesm values

* some variables should be used from the shel inp routine for
ufs but set locally for cesm
* restarts, grd and point output being written for ufs but
file name has no time string information

DeniseWorthen committed:
tidy up

DeniseWorthen committed:
sandbox testing for ww3 restart repro

*in debug mode, using dedicated ('second') output stream
produces non-identical ww3 restarts, but the model itself reproduces
this mode produces timestamped restart files

* in debug mode, using no data server mode, the ww3 restarts are
reproducible. this mode produces restarts restartXXX.ww3. base case
produces restarts at 09,012 (check in log.ww3), filenames restart001,
restart002. rest case produces only h=012, filename restart001.
cmp rest/restart001.ww3 base/restart002.ww3  shows the ww3 restarts
are b4b

DeniseWorthen committed:
fix time block for cesm

* move 'set time frame' block earlier to allow cesm
access to time and dtime_sync to set odat
* force odat(j=4) to be printed only once to log in
wav_shel_inp
* added "1" to end of point and grid out put spec in
ww3_shel.inp input file to force pre-pending by time

DeniseWorthen committed:
move ufs calls and make one cesmcoupled block

DeniseWorthen committed:
Merge branch 'feature/w3capbs3' into feature/ww3capbuild

DeniseWorthen committed:
remove scrip switches

DeniseWorthen committed:
fix two uninit faults using -init=snan,arrays

DeniseWorthen committed:
test branch for MV's latest changes

DeniseWorthen committed:
remove debugging code

DeniseWorthen committed:
remove more debugging code

DeniseWorthen committed:
update w/ hafs changes

DeniseWorthen committed:
update wav_comp_nuopc

DeniseWorthen committed:
clean up dbug statements and subnames

DeniseWorthen committed:
baselines pass with bs4 ww3 branch (c40a2762)

DeniseWorthen committed:
updates for merge_import

* remove flds_scalar_name from ToWave list, simplifies logic for
check_global SR
* set import_mask at start of import_fields using global reduction
and uwnd field from import state
* adds check_global routine for dumping out fields read from file,
the merged fields when merge_import is true and the import_mask

DeniseWorthen committed:
Merge branch 'feature/w3capbs4' of https://github.com/DeniseWorthen/WW3 into feature/w3capbs4

DeniseWorthen committed:
call set_importmask only for merge_import true

DeniseWorthen committed:
add comment, specify mdsf = ndsf(3)

* in ww3_shel.F90, ndsf(3) is assigned unit number 13 so use
that unit number in readfromfile

DeniseWorthen committed:
increase debug flag setting

* set dbug_flag to > 10 to write check_global

DeniseWorthen committed:
bring in MV's latest changes

DeniseWorthen committed:
compile fix

* remove import_mask from argument list

Commit:
added new cesmcoupled ifdefs to a subset of needed files

Commit:
updates for all CESMCOUPLED ifdefs

Commit:
fixed compile bug for ufs

Commit:
updates to get things running

DeniseWorthen committed:
updates to build and compile fix

* remove wav_wrapper_mod from build
* fix compile error

Commit:
updates to get cesm cases working out of the box

Commit:
updates to remove ww3.log file

Commit:
removed CESMCOUPLED if-defs that are no longer needed

Commit:
change needed for ufs build

Commit:
first set of changes to get debug mode working

DeniseWorthen committed:
add mesh mask check feature

* checks if mask in the provided mesh is all 1s (so
no valid land/sea mask) and replaces mesh mask with
internal land/sea mask

Commit:
fixed problems related to history writes

DeniseWorthen committed:
remove check for elementmask in mesh

mesh will contain an elementmask always, so don't need
to check if one exists

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/ww3capbuild

DeniseWorthen committed:
Merge remote-tracking branch 'cesm/dev/unified' into feature/ww3capbuild

DeniseWorthen committed:
update bin and inp for ufs

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/ww3capbuild

DeniseWorthen committed:
add SetVm and change log write from error to info

alperaltuntas committed:
add ww3a grid and IC files

alperaltuntas committed:
remove inadvertendly added __pycache__/ dir

alperaltuntas committed:
Merge pull request #15 from ESCOMP/add_ww3a

Add ww3a grid to dev/unified

DeniseWorthen committed:
modify setglobalinput

* do reduce across only sea points

DeniseWorthen committed:
modifications for nsea global reduce

* do reduce across only sea points
* add interface to do fill of global ww3 fields
* change variable name from data_global to global_data
* use ungridded dimension for wave elevation spectrum
* move util routines state_getfldptr and state_fldchk from wav_import_export to wav_shr mod

DeniseWorthen committed:
add interface to do fill of global ww3 fields

DeniseWorthen committed:
make merge_import work with nsea points

DeniseWorthen committed:
change variable name

rename data_global to global_data

DeniseWorthen committed:
clean up wav_coupling_to_cice arrays

* use ungridded dimension for wave_elevation_spectrum

DeniseWorthen committed:
compile fixes; move two SRS to wav_shr_mod

* move utility routines state_getldptr and state_fldchk from wav_import_export
to wav_shr_mod

DeniseWorthen committed:
cleanup wave_elevation_spectrum in wav_comp_nuopc

DeniseWorthen committed:
create multigrid branch of globalnsea

* add features from mcapstash

DeniseWorthen committed:
fix missing build changes; add comment

DeniseWorthen committed:
add missing multigrid conditional

DeniseWorthen committed:
move iounit setup for shel

* cleans up wav_comp_nuopc by creating new sr for opening
io units when not multigrid

DeniseWorthen committed:
add switch UWM

* add switch UWM for resetting charn in w3updtdmd for restart
repro

DeniseWorthen committed:
add switch and w3updtmd for uwm changes

DeniseWorthen committed:
add uwm netcdf gridded write changes

* optionally write netcdf grd output for uwm
* if history configuration is not set, will write native binary
ww3 gridded output

DeniseWorthen committed:
fix two compile errors

DeniseWorthen committed:
add missing time initialization for cesm

for multigrid, time variable cannot be set until after
initialization. Error is Attempt to use pointer TIME when it is not associated with a target
ufs sets time from time0 in read_shel_inp but cesm time is used to set odat so it must
be set prior to calling waveinit_cesm

Commit:
new #ifdefs introduced to not change the build for cesm

DeniseWorthen committed:
fix compile error

* all uwm baselines pass for both shel and multigrid

DeniseWorthen committed:
Merge remote-tracking branch 'origin/feature/globalnsea_multi' into feature/globalnsea_gridnc

DeniseWorthen committed:
revert uwm changes in w3wavemd

DeniseWorthen committed:
return uwm ifdef option to w3wavemd

DeniseWorthen committed:
initial commit for refac iogoncmd

DeniseWorthen committed:
remove cime_config and CESM CMakeLists.txt

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/cmake_meshcapbuild

DeniseWorthen committed:
revert switch file and add swith_meshcap

kgerheiser committed:
Move switches to bin

kgerheiser committed:
Add option for multiple caps through the UFS_CAP option

Supported values are MULTI_ESMF and NUOPC_MESH

kgerheiser committed:
Test UFS_CAP options in workflows

DeniseWorthen committed:
Merge pull request #2 from kgerheiser/feature/cmake_meshcapbuild

Add new cap options to CMake

DeniseWorthen committed:
fix errors

DeniseWorthen committed:
initial commit of refactored iogonc

DeniseWorthen committed:
commit initial working code

* currently writing 2 test fields correctly

DeniseWorthen committed:
fix reversed logic for cap type

DeniseWorthen committed:
remove procedure interface

* call appropriate write_var_X SR for each varying axis
* add netcdf handle ierr call

DeniseWorthen committed:
fill out remaining variables

* add additional comments

DeniseWorthen committed:
fix compile errors

DeniseWorthen committed:
prepare merge w/ ufs cmake build for mesh cap

* remove cime_config and cesm CMakelist from branch committed
to dev/unified in order to merge cleanly with cmake mesh cap
changes on ufs side

DeniseWorthen committed:
Merge remote-tracking branch 'origin/feature/no_uwm_ncgrid' into feature/cmake_meshcapbuild

DeniseWorthen committed:
compile and grd.out fixes

* fix github action compile failure
* create new switch setting specifically to allow uwm netcdf
output, otherwise native binary output will be written. This is not
as flexible as initially implmented, which allowed netcdf gridded
output as a config flag. We reverted that ability in w3iogoncmd to
debug the perf issues on cesm side. UWM netcdf griddded output can
be turned on with compile-time switch UWMNCOUT
* all baselines pass run against shel_wavebl_20220214

DeniseWorthen committed:
remove cesm/cime related files

DeniseWorthen committed:
Merge branch 'feature/cmake_meshcapbuild' into feature/refac_iognc

DeniseWorthen committed:
add file to src list

DeniseWorthen committed:
fix compile error

alperaltuntas committed:
Merge pull request #18 from DeniseWorthen/feature/no_uwm_ncgrid

Implement use of NSEA points and time-axis in gridded output

DeniseWorthen committed:
add fix to allow compile with -DDEBUG=ON

* native instances of "DEBUG" were pre-processed (?) into "1"
so renamed to wDEBUG as temp workaround

DeniseWorthen committed:
intermediate code

* need to test against original ionc
* need to test moving initialization into write_var routines

kgerheiser committed:
Remove debug definition from UFS

A variable named DEBUG conflicts and is replaced by the value `1`.

DeniseWorthen committed:
Merge pull request #3 from kgerheiser/patch-1

Remove debug definition from UFS

Commit:
fixed dt settings so that they would not be overwritten by call to w3init

alperaltuntas committed:
Merge pull request #20 from mvertens/feature/fix_dtsettings

fixed dt settings so that they would not be overwritten by call to w3…

DeniseWorthen committed:
use interface for state_getfldptr utility routine

DeniseWorthen committed:
remove use statements for ESMF and NUOPC

* use ESMF and use NUOPC mean the 'only' lines are redundant

DeniseWorthen committed:
update cesm timestep setting

* fix dt initialization for cesm

DeniseWorthen committed:
Merge remote-tracking branch 'remotes/ww3-cesm/dev/unified' into feature/cmake_meshcapbuild

DeniseWorthen committed:
remove cesm section

DeniseWorthen committed:
Merge remote-tracking branch 'origin/feature/fix2d' into feature/cmake_meshcapbuild

alperaltuntas committed:
Add the gx1v7 grid input files

DeniseWorthen committed:
Merge remote-tracking branch 'origin/feature/cmake_meshcapbuild' into feature/refac_iognc

DeniseWorthen committed:
revert temp changes associated with DDEBUG

DeniseWorthen committed:
fix differences w/ mediator history files

DeniseWorthen committed:
add doxygen formatting to mesh cap code

Squashed commit of the following:

commit 444212dd945f1d29135514723d5b62be38542863
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Thu Apr 7 09:28:31 2022 -0400

    tidy up

commit 03435033b2aadc8b03791ef78bedbef2db5a5047
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Thu Apr 7 09:05:42 2022 -0400

    more doxygen formatting added

commit 61c10acdf81b2718bd90717bb83560e236aa9a9c
Author: denise.worthen <Denise.Worthen@noaa.gov>
Date:   Wed Apr 6 22:17:19 2022 +0000

    updates more doxy

commit c9d10e4b851f1ffebece694db3de292ea9133e2d
Author: denise.worthen <Denise.Worthen@noaa.gov>
Date:   Wed Apr 6 22:02:51 2022 +0000

    fix doxy errors

    * still have warning that u_file_u is not documented ?

commit d19652f526f208ab23deee9266a037573cc75090
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Wed Apr 6 17:15:34 2022 -0400

    more doxy

commit f1049d01d987fd325826c132ae2f37802dbd908e
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Wed Apr 6 08:22:54 2022 -0400

    minor text placement changes

commit 24f99451b767c5fdc5a84fcb86610965963485ed
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Tue Apr 5 15:51:47 2022 -0400

    update 2 more files

commit fe25b6321e3b2063cda4aa721f8eb6d5c31c08be
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Tue Apr 5 12:58:03 2022 -0400

    initial commit of added doxygen

DeniseWorthen committed:
fix b axis array and white space cleanup

DeniseWorthen committed:
replace routines when variable has extra dim

* use a single write_var3d routine to write any variable
with an extra axis
* b4b w/ output from previous method (both in debug mode)

DeniseWorthen committed:
Merge branch 'feature/cmake_meshcapbuild' into feature/refac_iognc

DeniseWorthen committed:
passes comparison w/ original iogoncmd code

* ww3 gridded out files are b4b with original code
* mediator history files for wave are b4b w/ original code
* both tests run in debug mode

DeniseWorthen committed:
fix time variable

DeniseWorthen committed:
refactor initializtion

* remove most of initialization and place w/in the actual write
routine
* ww3 output is b4b w/ compared to the previous method; tested
with writing out all possible variables w/ the exception of the
freq variables (needs a mod_def change)

DeniseWorthen committed:
remove remaining init section, remove fname from call

* add init2 optional input to initialize subset of variables
where mapsta==2
* remove fname from calls to write_var2d/3d
* change case on handle_err function

DeniseWorthen committed:
remove need for allocate/deall for 3d-variables

* allocate 3d var only if a requested output variable
needs it; define these as targets and point to them w/in the
write_var3d routine
* reorg the writing of variables into groups for each extra
axis required
* remove 'x y' from dims listing--these are default for gridded
output
* all outputs are b4b w/ outputs prior to refactoring initialization
and switch to single var3d routine

DeniseWorthen committed:
remove flogrd from iogonc routine

DeniseWorthen committed:
replace use of scan for additional dims

DeniseWorthen committed:
correct many missing nd units

* use w3ounfmetamd.F90 meta type to add correct units

DeniseWorthen committed:
remove 'b' axis

* bedforms split into 3 nsea arrays based on w3ounfmetamd.F90

DeniseWorthen committed:
change format, clean up

* remove some remaining b-axis variables

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/cmake_meshcapbuild

alperaltuntas committed:
Merge pull request #21 from ESCOMP/add_gx1v7_to_dev/unified

Add the gx1v7 grid input files

DeniseWorthen committed:
address pr review comments

* change cesmcoupled to w3_cesmcoupled
* removed commented out line in w3init
* mark sections which require fixes and testing
for cesm

DeniseWorthen committed:
add custom user restart and history filenames

DeniseWorthen committed:
address pr comments

* remove marked sections for pr review

DeniseWorthen committed:
remove debugging statements for CESMCOUPLED

DeniseWorthen committed:
fix compile error

alperaltuntas committed:
remove cime_config from core WW3

DeniseWorthen committed:
switch to ncout for ufs, add ptr restart example

DeniseWorthen committed:
updates for initfile

* if runtype=initial, user_restfname will be set==initfile
* if reading restart, file will be either initfile (initial
runtype) or the full restart filename

alperaltuntas committed:
standardize format. @jedwards4b

alperaltuntas committed:
Merge pull request #1 from ESCOMP/equation_format_fix

standardize format

Commit:
updates to changes from dworthen that fix the wave spectrum fields sent to the mediator

DeniseWorthen committed:
Merge branch 'feature/cmake_meshcapbuild' into feature/refac_iognc

DeniseWorthen committed:
Merge branch 'feature/cmake_meshcapbuild' into feature/useropts

Commit:
added wav_grdout.F90

DeniseWorthen committed:
fix issues w/ setting restart file name

* add default name for gridded out netcdf file (not yet tested)

DeniseWorthen committed:
isolate use file name options

* github actions fail when using options from wav_shr_mod

alperaltuntas committed:
Merge pull request #2 from mvertens/feature/icewav_coupling

updates to ww3 cap code for ww3/cice coupling

DeniseWorthen committed:
clean up, move runtype and initfile to w3odat

* remove creating of history file name in w3iogoncmd
* move runtype and initfile to w3odat so they can be
used in w3iorsmd w/o an ifdef

alperaltuntas committed:
merge with cmake_meshcapbuild and resolve conflicts

alperaltuntas committed:
define __WW3_SWITCHES__ in CMakeLists

alperaltuntas committed:
Merge pull request #3 from ESCOMP/merge_cmake_meshcapbuild

Merge mesh cap changes from (PR #641)

alperaltuntas committed:
comment out wav_tauice export lines

DeniseWorthen committed:
fix requests of unallocated variables

* remove requested variables if the right mod_def settings are
not used; these all seem to require OUTS E3D=1 etc

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/useropts

DeniseWorthen committed:
add convert statements

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/refac_iognc

DeniseWorthen committed:
fix math for sic4, minor logging changes

* add changes from escomp for calc karg1 in sic4
* add message about current log file and init some strings

DeniseWorthen committed:
logging changes

* remove mod(hh,outfreq); when outfreq=0, this hangs if uncommented

DeniseWorthen committed:
update w3iogomd for optional filenames

* update binary gridded output for user selected
filenames. b4b w/ current when binary output is used

DeniseWorthen committed:
remove unused variables from w3iogomd

DeniseWorthen committed:
initial commit of useropts + alarms

* test if b4b with useropts+uwmncout is used

DeniseWorthen committed:
fix double (binary,nc) history file write

* fix ci (?) by moving iogonc to main src list

DeniseWorthen committed:
add options via config

* using user names for hist and rest are now read from config
* gridded netcdf is read from config
* for cesm, set user names and gridded as default in cesm init
* allow gridded netcdf to be written at inp frequency

DeniseWorthen committed:
initial attempt for full inp/nml functionality

DeniseWorthen committed:
fix compile errors, indenting on shel_inp

DeniseWorthen committed:
remove reading of input files

DeniseWorthen committed:
add only to use w3nmshelmd, add flags

Commit:
updates to be able to remove wav_shel_set routine in wav_shel_inp

Commit:
reformatted wav_shel_inp.F90

DeniseWorthen committed:
switch to using interface for debug print msgs

DeniseWorthen committed:
manually bring in changes from MV

*change name of read_shel_inp to read_shel_config since
it reads either inp or nml file
* change name of debuginit print utility because it will be
general to multiple debug options
* remove set_shel_inp for cesm and instead call read_shel_config
* make formatting changes in read_shel_config
* changes brought in manually from MV/feature/wav_shel_inp_namelist

Commit:
added logic for turning on wave/ice coupling

Commit:
moved CESMCOUPLED to W3_CESMCOUPLED

DeniseWorthen committed:
fix all caps line and some doxy formatting

DeniseWorthen committed:
remove file committed by error

DeniseWorthen committed:
add mv's changes to rename flags w/ w3_ prefix

Commit:
updated all wav_*.F90 files to be consistent with latest changes from Denise Worthen

Commit:
Merge branch 'dev/unified' into feature/wav_shel_inp_namelist

Commit:
fixed compile problems

Commit:
merged to head of denise/feature/playwithshel

DeniseWorthen committed:
Merge branch 'feature/useropts' into feature/playwithshel

DeniseWorthen committed:
fix wav_ice coupling flag and remove tauice

Commit:
merged to user output file PR from Denise

Commit:
updates to get branch working and debug output working

Commit:
Merge branch 'feature/denise_output' into feature/wav_shel_inp_namelist

DeniseWorthen committed:
Merge branch 'develop' into feature/update_to_develop

DeniseWorthen committed:
Merge remote-tracking branch 'origin/feature/update_to_develop' into feature/refac_iognc

DeniseWorthen committed:
add all switch flags, remove initialization

* remove need to call initialize for flags (switches)
* add all switches defined in manual
* add 'only' to all use statements in w3wavemd and w3iorsmd

DeniseWorthen committed:
add flags for cesm and uwm

DeniseWorthen committed:
Merge branch 'feature/playwithshel' into feature/refac_iognc

DeniseWorthen committed:
comment out dbug_flag (need to move from wav_shr)

DeniseWorthen committed:
move wav_gridout compile to main list

DeniseWorthen committed:
fix pet logging issue when using nml, fix shr_flag

* For point output, use open(newunit=ndsl....). Previously, ndsl
was assigned to unit=50, which corrupts the PET logging function
(pet logging would be switched to fort.50).
* fix bad copy/paste error where cesm flag was missing type (logical)

DeniseWorthen committed:
update w/ MV's latest changes

* change variable names for user set items
* add a couple more flags, remove ifdefs on formats

DeniseWorthen committed:
fix writing of odat to fort file when shel is used

DeniseWorthen committed:
Merge remote-tracking branch 'origin/feature/playwithshel' into feature/refac_iognc

DeniseWorthen committed:
update shr_flags and fix white space

* add all defined flags to wav_shr_flags
* make all flags available in wav_shel_inp.F90
* add print_memcheck utility
* clean up white space w3meminfo.F90

DeniseWorthen committed:
Merge branch 'feature/playwithshel' into feature/refac_iognc

DeniseWorthen committed:
change case for iogonc, fix cesm ifdef

DeniseWorthen committed:
all wave tests are b4b with uwm hash b1a6a936

* baseline 20220629

Commit:
merged to Denise Worthen's branch feature/refac_iognc

Commit:
fixes for restart test

DeniseWorthen committed:
Merge branch 'feature/refac_iognc' into feature/playwithshel

DeniseWorthen committed:
Merge remote-tracking branch 'mvertens/feature/refac_iognc_mvertens' into feature/playwithshel

DeniseWorthen committed:
minor cleanup of wav_shel_inp

* add msgunit = 740+iaproc for either memcheck or debugint
* replace W3_MEMCHECK with call to print_memcheck

DeniseWorthen committed:
text alignment

DeniseWorthen committed:
align text in w3iogonc and wav_shr_flags

DeniseWorthen committed:
update to remove histalarm and restalarm options

* use ringinterval set from configuration attributes
to set dtout(1) for history and dtout(4 or 8) for
restarts. this allows removal of user_restalarm and
user_histalarm. gridded output and restarts can be
written in w3wave using the native ww3 timer.
* rename user_grdncout to user_netcdf_grdout
* remove outfreq since it is not used
* text alignment and other whitespace changes

DeniseWorthen committed:
fix open unit, log file message

* close unit ndsi when reading shel.inp file
* write log message for naprst in w3iorsmd

DeniseWorthen committed:
restor histwr and restwr to w3odatmd

* cesm needs rstwr control to prevent writing restart
at initial time

DeniseWorthen committed:
fix endless loop for cesm style dates

* use timen,time0 to set odat in wav_shel_inp

DeniseWorthen committed:
use histwr for w3iogoncd when cesmcoupled

DeniseWorthen committed:
try compile fix

DeniseWorthen committed:
Revert "try compile fix"

This reverts commit 1bc5a86f6fea711ee97ebe09e72e0e49d42ec638.

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/playwithshel

DeniseWorthen committed:
revert changes which broke cesm; add flag control

* revert alarm control which broke cesm restart and history writes
* add initial flag control for debug and tracing options in w3wave and w3init primarily
* move implicit none to module default
* remove serial ifdefs which appear once flags control some options
* indent and text align

JessicaMeixner-NOAA commented:
@DeniseWorthen and @mvertens  -- WOW! So much great work!   I've already started looking through this and look forward to working with you with the restart naming conventions to expand this capability to be a feature usable by standalone WW3 and not just a NUOPC coupled application.    

I have some concerns as I look through this: 
1. We have a lot of unrelated work on many topics, when it's preferred to split coherent pieces of work into separate branches/PR when possible. It really seems like this could be a few smaller PRs, but perhaps I do not yet understand the interconnectedness. 
2. There are a lot of updates to the code (Thank you so much for all of the additional only in module use statements!) that seem unrelated to these updates describe in the PR text and will actually diverge this branch further from develop and make it harder to merge it back into develop.  Perhaps I can help with some of these updates by seperating them out and putting them through a separate review and test process in the develop branch? 
3. There are some whole-scale style changes such as the #ifdef debug that should really go through it's own review process to make sure the whole WW3 community is on board with some of these changes.   @aliabdolali have you scheduled that talk yet?  I know you were aiming for September.  
 
I want to emphasize that I personally am really excited about all of these updates and am happy to work with you both to make sure everything gets included and this branch gets back to the develop branch as soon as possible.

DeniseWorthen committed:
minor text fixes

DeniseWorthen commented:
This PR resolves several issues related to getting the mesh cap into develop. The changes are not in fact un-related; they all revolve around file in/out (i.e. using a ww3_shel.nml for CESM, user file names and gridded netcdf output). Keeping the changes together means that the mesh cap can come to develop sooner. It has been extensively tested during development and confirmed to both work for CESM and have no impact on UWM. 

In terms of converting ifdefs, since @mvertens and I are doing this work in the mesh cap, there is no way for us to make the changes and also test unless we make the changes in that mesh cap branch. While the text changes are extensive, as seen in this [comparison](https://github.com/NOAA-EMC/WW3/pull/757/files?w=1), they specifically touch only a small subset of ifdefs or variable declarations. However, even this limited application we believe demonstrates the advantages in code clarity and readability of moving down this path. 

I do have a branch which removes all the mesh cap items but contains the same ifdef changes. I can PR that to develop in order to keep the two branches in sync and perhaps spark the discussion you reference. 

However, if the removal of ifdefs and a transition to run-time control of options is desired goal, the work @mvertens and I have done at this point make clear to us that there is no way to do that and not be disruptive.

JessicaMeixner-NOAA commented:
@DeniseWorthen submitting a PR with just the ifdef changes to the develop branch would be great.  That will surely start the discussion.  Converting the if-defs is not required for the mesh cap to be merged back to the develop branch.  In general,  If there's a disruptive change coming, it's preferred that it's in a focused PR (to the develop branch) that is not combined with other updates which helps with code management and future debugging.

DeniseWorthen committed:
add two small changes consistent with develop

DeniseWorthen committed:
fix recv_init for langmt

DeniseWorthen committed:
resolve circular dependency, fix text

DeniseWorthen committed:
changes related to wave-ice coupling

* fix name of config variable to enable wave-cice coupling
* in w3sic4, remove cesm ifdefs and icecon variable which is used
for debug message only
* for uwm, need to switch ic0->ic4 in meshcap switch to enable and
use mod_def with &SIC4 IC4METHOD = 8 and enable freq output variable
&OUTS E3D = 1, USSP = 1, IUSSP = 3, STK_WN = 0.04, 0.110, 0.3305

DeniseWorthen committed:
make flags into parameters

* allows compiler to optimize code w/in if-flag blocks

DeniseWorthen committed:
fix rev in for langmt

DeniseWorthen committed:
update workflow for mesh cap build

DeniseWorthen committed:
initialize variable XXX in w3wavemd

DeniseWorthen committed:
move some implict none to module level

* text cleanup

DeniseWorthen committed:
fix typo

DeniseWorthen committed:
fix broadcast of initfile, minor text cleanup
